File: sdk/typescript-client/src/management.ts
export type AuditAckStatus = number;
export const AuditAckStatus = {
  Unspecified: 0,
  Durable: 1,
  OutcomeUnknown: 2,
  AuditUnavailable: 3,
  Disabled: 4,
} as const;

export type CancelDisposition = number;
export const CancelDisposition = {
  Unspecified: 0,
  Accepted: 1,
  AlreadyTerminal: 2,
  NotFound: 3,
} as const;

export type CapabilityPolicyRecordKind = number;
export const CapabilityPolicyRecordKind = {
  Unspecified: 0,
  Policy: 1,
  ProviderBinding: 2,
} as const;

export type FailureCategory = number;
export const FailureCategory = {
  Unspecified: 0,
  LocalCancelled: 1,
  Deadline: 2,
  Transport: 3,
  Rpc: 4,
  Decode: 5,
  Limit: 6,
  InvalidRequest: 7,
} as const;

export type OutcomeKnowledge = number;
export const OutcomeKnowledge = {
  Unspecified: 0,
  NotDispatched: 1,
  Unknown: 2,
  Observed: 3,
} as const;

export interface ResourceBudget {
  readonly cpuFuel: bigint;
  readonly memoryBytes: bigint;
  readonly childCalls: number;
  readonly outboundRequests: number;
  readonly stateReadBytes: bigint;
  readonly stateWriteBytes: bigint;
  readonly blobReadBytes: bigint;
  readonly blobWriteBytes: bigint;
  readonly logBytes: bigint;
  readonly effectCount: number;
  readonly wallTimeLimitMillis?: bigint;
}

export interface ErrorDetail {
  readonly kind: string;
  readonly fields: Readonly<Record<string, string>>;
}

export interface PlatformError {
  readonly code: string;
  readonly message: string;
  readonly retryable: boolean;
  readonly detailItems: readonly ErrorDetail[];
}

export interface ObjectMetadata {
  readonly name: string;
  readonly tenant?: string;
  readonly namespace?: string;
  readonly labels: Readonly<Record<string, string>>;
  readonly annotations: Readonly<Record<string, string>>;
}

export interface PageRequest {
  readonly pageSize: number;
  readonly pageToken?: string;
}

export interface PageResponse {
  readonly nextPageToken?: string;
}

export interface AuditAck {
  readonly status: AuditAckStatus;
  readonly attemptSequence?: bigint;
}

export interface InvocationTarget {
  readonly tenant: string;
  readonly service: string;
  readonly contract: string;
  readonly function: string;
  readonly route?: string;
}

export interface InvokeRequest {
  readonly activationId?: string;
  readonly parentActivationId?: string;
  readonly rootActivationId?: string;
  readonly target?: InvocationTarget;
  readonly payload: Uint8Array;
  readonly mediaType: string;
  readonly deadlineUnixMillis?: bigint;
  readonly priority: number;
  readonly idempotencyKey?: string;
  readonly budget?: ResourceBudget;
  readonly metadata: Readonly<Record<string, string>>;
}

export interface BudgetConsumption {
  readonly cpuFuel: bigint;
  readonly peakMemoryBytes: bigint;
  readonly wallTimeMicros: bigint;
  readonly childCalls: number;
  readonly outboundRequests: number;
  readonly stateReadBytes: bigint;
  readonly stateWriteBytes: bigint;
  readonly blobReadBytes: bigint;
  readonly blobWriteBytes: bigint;
  readonly logBytes: bigint;
  readonly effectCount: number;
}

export interface Success {
  readonly payload: Uint8Array;
  readonly mediaType: string;
  readonly committedStateVersion?: string;
  readonly effectIds: readonly string[];
  readonly metadata: Readonly<Record<string, string>>;
}

export interface DeclaredError {
  readonly code: string;
  readonly message: string;
  readonly payload: Uint8Array;
  readonly mediaType: string;
  readonly metadata: Readonly<Record<string, string>>;
}

export interface InvokeResponse {
  readonly activationId: string;
  readonly revisionId: string;
  readonly releaseDigest: string;
  readonly routeGeneration: bigint;
  readonly success?: Success;
  readonly declaredError?: DeclaredError;
  readonly platformFailure?: PlatformError;
  readonly consumption?: BudgetConsumption;
  readonly publicationId?: string;
}

export interface CancelRequest {
  readonly activationId: string;
  readonly reason: string;
}

export interface CancelResponse {
  readonly disposition: CancelDisposition;
  readonly terminalState?: string;
}

export interface GetActivationRequest {
  readonly activationId: string;
}

export interface ActivationSuccessSummary {
  readonly committedStateVersion?: string;
  readonly effectIds: readonly string[];
  readonly metadata: Readonly<Record<string, string>>;
}

export interface ActivationStatus {
  readonly activationId: string;
  readonly phase: string;
  readonly terminalState?: string;
  readonly lastUpdatedUnixMillis: bigint;
  readonly metadata: Readonly<Record<string, string>>;
  readonly succeeded?: ActivationSuccessSummary;
  readonly declaredError?: DeclaredError;
  readonly platformFailure?: PlatformError;
  readonly finalConsumption?: BudgetConsumption;
  readonly terminalAtUnixMillis?: bigint;
}

export interface Policy {
  readonly id: string;
  readonly metadata?: ObjectMetadata;
  readonly document: string;
  readonly generation: bigint;
  readonly language: string;
  readonly recordKind: CapabilityPolicyRecordKind;
  readonly contentDigest: string;
  readonly revoked: boolean;
}

export interface ApplyPolicyRequest {
  readonly policy?: Policy;
  readonly expectedGeneration?: bigint;
  readonly operationId: string;
}

export interface CapabilityPolicyOperation {
  readonly operationId: string;
  readonly tenant: string;
  readonly id: string;
  readonly recordKind: CapabilityPolicyRecordKind;
  readonly generation: bigint;
  readonly contentDigest: string;
  readonly revoked: boolean;
}

export interface ApplyPolicyResponse {
  readonly policy?: Policy;
  readonly receipt?: CapabilityPolicyOperation;
}

export interface GetPolicyRequest {
  readonly id: string;
  readonly recordKind: CapabilityPolicyRecordKind;
}

export interface GetPolicyResponse {
  readonly policy?: Policy;
}

export interface GetPolicyOperationRequest {
  readonly operationId: string;
}

export interface GetPolicyOperationResponse {
  readonly receipt?: CapabilityPolicyOperation;
}

export interface ListPoliciesRequest {
  readonly recordKind: CapabilityPolicyRecordKind;
  readonly page?: PageRequest;
}

export interface ListPoliciesResponse {
  readonly policies: readonly Policy[];
  readonly catalogGeneration: bigint;
  readonly page?: PageResponse;
}

export interface CapabilityInspectionPolicy {
  readonly id: string;
  readonly revision: bigint;
  readonly digest: string;
}

export interface CapabilityBindingInspection {
  readonly definitionDigest?: string;
  readonly providerBinding?: CapabilityInspectionPolicy;
  readonly policies: readonly CapabilityInspectionPolicy[];
  readonly providerProfile: string;
  readonly providerConfigurationDigest: string;
  readonly providerConfigurationEpoch: bigint;
  readonly state: string;
}

export interface CapabilityDescriptor {
  readonly id: string;
  readonly contract: string;
  readonly provider: string;
  readonly operations: readonly string[];
  readonly attributes: Readonly<Record<string, string>>;
  readonly inspection?: CapabilityBindingInspection;
}

export interface ListCapabilitiesRequest {
  readonly contractPrefix?: string;
  readonly provider?: string;
  readonly page?: PageRequest;
  readonly deploymentId: string;
  readonly includeNodeUsage: boolean;
}

export interface CapabilityInspectionRevision {
  readonly deploymentId: string;
  readonly revisionId: string;
  readonly componentDigest: string;
  readonly publicationId?: string;
  readonly routeGeneration: bigint;
  readonly catalogTransaction: bigint;
}

export interface CapabilityResourceUsage {
  readonly scope: string;
  readonly counters: Readonly<Record<string, bigint>>;
  readonly unavailable: readonly string[];
}

export interface ListCapabilitiesResponse {
  readonly capabilities: readonly CapabilityDescriptor[];
  readonly page?: PageResponse;
  readonly revision?: CapabilityInspectionRevision;
  readonly tenantUsage?: CapabilityResourceUsage;
  readonly nodeUsage?: CapabilityResourceUsage;
  readonly state: string;
}

export interface CapabilityInspectionCeiling {
  readonly operations: number;
  readonly inputBytes: bigint;
  readonly outputBytes: bigint;
  readonly wallTimeMillis: bigint;
}

export interface PublicationRef {
  readonly id: string;
  readonly tenant: string;
}

export interface ReleaseSelector {
  readonly componentDigest?: string;
  readonly publication?: PublicationRef;
}

export interface PublicationIdentity {
  readonly publication: PublicationRef;
  readonly componentDigest: string;
  readonly packageDigest: string;
}

export interface CallOptions {
  readonly timeoutMillis?: bigint;
  readonly signal?: AbortSignal;
}

export interface RequestIdentity {
  readonly activationId?: string;
  readonly operationId?: string;
}

export interface ResponseMetadata {
  readonly identity: RequestIdentity;
  readonly outcome: OutcomeKnowledge;
  readonly auditAck?: AuditAck;
  readonly auditStatus?: string;
}

export interface ClientFailure {
  readonly category: FailureCategory;
  readonly message: string;
  readonly grpcStatus?: number;
  readonly platformError?: PlatformError;
  readonly dispatched: boolean;
  readonly outcome: OutcomeKnowledge;
  readonly identity: RequestIdentity;
  readonly auditAck?: AuditAck;
  readonly auditStatus?: string;
}

export interface ClientResponse<Response> {
  readonly value: Response;
  readonly metadata: ResponseMetadata;
}

export interface ClientProfile {
  invoke(request: InvokeRequest, options?: CallOptions): Promise<ClientResponse<InvokeResponse>>;
  cancel(request: CancelRequest, options?: CallOptions): Promise<ClientResponse<CancelResponse>>;
  getActivation(request: GetActivationRequest, options?: CallOptions): Promise<ClientResponse<ActivationStatus>>;
  getPolicy(request: GetPolicyRequest, options?: CallOptions): Promise<ClientResponse<GetPolicyResponse>>;
  listPolicies(request: ListPoliciesRequest, options?: CallOptions): Promise<ClientResponse<ListPoliciesResponse>>;
  listCapabilities(request: ListCapabilitiesRequest, options?: CallOptions): Promise<ClientResponse<ListCapabilitiesResponse>>;
  applyPolicy(request: ApplyPolicyRequest, options?: CallOptions): Promise<ClientResponse<ApplyPolicyResponse>>;
  getPolicyOperation(request: GetPolicyOperationRequest, options?: CallOptions): Promise<ClientResponse<GetPolicyOperationResponse>>;
}

export class ClientError extends Error {
  constructor(readonly failure: ClientFailure) {
    super(failure.message);
    this.name = "ClientError";
  }
}

export function parseU64Decimal(value: string): bigint {
  if (typeof value !== "string" || !/^(0|[1-9][0-9]{0,19})$/.test(value)) {
    throw new RangeError("invalid uint64 decimal");
  }
  const parsed = BigInt(value);
  if (parsed > 18446744073709551615n) throw new RangeError("uint64 overflow");
  return parsed;
}

export function formatU64Decimal(value: bigint): string {
  if (typeof value !== "bigint" || value < 0n || value > 18446744073709551615n) {
    throw new RangeError("invalid uint64 bigint");
  }
  return value.toString(10);
}

File: sdk/java-client/src/main/java/dev/latent/sdk/Management.java
package dev.latent.sdk;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public final class Management {
    private Management() { }

    public record AuditAckStatus(int value) {
        public static final AuditAckStatus UNSPECIFIED = new AuditAckStatus(0);
        public static final AuditAckStatus DURABLE = new AuditAckStatus(1);
        public static final AuditAckStatus OUTCOME_UNKNOWN = new AuditAckStatus(2);
        public static final AuditAckStatus AUDIT_UNAVAILABLE = new AuditAckStatus(3);
        public static final AuditAckStatus DISABLED = new AuditAckStatus(4);
    }

    public record CancelDisposition(int value) {
        public static final CancelDisposition UNSPECIFIED = new CancelDisposition(0);
        public static final CancelDisposition ACCEPTED = new CancelDisposition(1);
        public static final CancelDisposition ALREADY_TERMINAL = new CancelDisposition(2);
        public static final CancelDisposition NOT_FOUND = new CancelDisposition(3);
    }

    public record CapabilityPolicyRecordKind(int value) {
        public static final CapabilityPolicyRecordKind UNSPECIFIED = new CapabilityPolicyRecordKind(0);
        public static final CapabilityPolicyRecordKind POLICY = new CapabilityPolicyRecordKind(1);
        public static final CapabilityPolicyRecordKind PROVIDER_BINDING = new CapabilityPolicyRecordKind(2);
    }

    public record FailureCategory(int value) {
        public static final FailureCategory UNSPECIFIED = new FailureCategory(0);
        public static final FailureCategory LOCAL_CANCELLED = new FailureCategory(1);
        public static final FailureCategory DEADLINE = new FailureCategory(2);
        public static final FailureCategory TRANSPORT = new FailureCategory(3);
        public static final FailureCategory RPC = new FailureCategory(4);
        public static final FailureCategory DECODE = new FailureCategory(5);
        public static final FailureCategory LIMIT = new FailureCategory(6);
        public static final FailureCategory INVALID_REQUEST = new FailureCategory(7);
    }

    public record OutcomeKnowledge(int value) {
        public static final OutcomeKnowledge UNSPECIFIED = new OutcomeKnowledge(0);
        public static final OutcomeKnowledge NOT_DISPATCHED = new OutcomeKnowledge(1);
        public static final OutcomeKnowledge UNKNOWN = new OutcomeKnowledge(2);
        public static final OutcomeKnowledge OBSERVED = new OutcomeKnowledge(3);
    }

    public record ResourceBudget(
            long cpuFuel,
            long memoryBytes,
            int childCalls,
            int outboundRequests,
            long stateReadBytes,
            long stateWriteBytes,
            long blobReadBytes,
            long blobWriteBytes,
            long logBytes,
            int effectCount,
            Optional<Long> wallTimeLimitMillis) { }

    public record ErrorDetail(
            String kind,
            Map<String, String> fields) { }

    public record PlatformError(
            String code,
            String message,
            boolean retryable,
            List<ErrorDetail> detailItems) { }

    public record ObjectMetadata(
            String name,
            Optional<String> tenant,
            Optional<String> namespace,
            Map<String, String> labels,
            Map<String, String> annotations) { }

    public record PageRequest(
            int pageSize,
            Optional<String> pageToken) { }

    public record PageResponse(
            Optional<String> nextPageToken) { }

    public record AuditAck(
            AuditAckStatus status,
            Optional<Long> attemptSequence) { }

    public record InvocationTarget(
            String tenant,
            String service,
            String contract,
            String function,
            Optional<String> route) { }

    public record InvokeRequest(
            Optional<String> activationId,
            Optional<String> parentActivationId,
            Optional<String> rootActivationId,
            Optional<InvocationTarget> target,
            ByteBuffer payload,
            String mediaType,
            Optional<Long> deadlineUnixMillis,
            int priority,
            Optional<String> idempotencyKey,
            Optional<ResourceBudget> budget,
            Map<String, String> metadata) { }

    public record BudgetConsumption(
            long cpuFuel,
            long peakMemoryBytes,
            long wallTimeMicros,
            int childCalls,
            int outboundRequests,
            long stateReadBytes,
            long stateWriteBytes,
            long blobReadBytes,
            long blobWriteBytes,
            long logBytes,
            int effectCount) { }

    public record Success(
            ByteBuffer payload,
            String mediaType,
            Optional<String> committedStateVersion,
            List<String> effectIds,
            Map<String, String> metadata) { }

    public record DeclaredError(
            String code,
            String message,
            ByteBuffer payload,
            String mediaType,
            Map<String, String> metadata) { }

    public record InvokeResponse(
            String activationId,
            String revisionId,
            String releaseDigest,
            long routeGeneration,
            Optional<Success> success,
            Optional<DeclaredError> declaredError,
            Optional<PlatformError> platformFailure,
            Optional<BudgetConsumption> consumption,
            Optional<String> publicationId) { }

    public record CancelRequest(
            String activationId,
            String reason) { }

    public record CancelResponse(
            CancelDisposition disposition,
            Optional<String> terminalState) { }

    public record GetActivationRequest(
            String activationId) { }

    public record ActivationSuccessSummary(
            Optional<String> committedStateVersion,
            List<String> effectIds,
            Map<String, String> metadata) { }

    public record ActivationStatus(
            String activationId,
            String phase,
            Optional<String> terminalState,
            long lastUpdatedUnixMillis,
            Map<String, String> metadata,
            Optional<ActivationSuccessSummary> succeeded,
            Optional<DeclaredError> declaredError,
            Optional<PlatformError> platformFailure,
            Optional<BudgetConsumption> finalConsumption,
            Optional<Long> terminalAtUnixMillis) { }

    public record Policy(
            String id,
            Optional<ObjectMetadata> metadata,
            String document,
            long generation,
            String language,
            CapabilityPolicyRecordKind recordKind,
            String contentDigest,
            boolean revoked) { }

    public record ApplyPolicyRequest(
            Optional<Policy> policy,
            Optional<Long> expectedGeneration,
            String operationId) { }

    public record CapabilityPolicyOperation(
            String operationId,
            String tenant,
            String id,
            CapabilityPolicyRecordKind recordKind,
            long generation,
            String contentDigest,
            boolean revoked) { }

    public record ApplyPolicyResponse(
            Optional<Policy> policy,
            Optional<CapabilityPolicyOperation> receipt) { }

    public record GetPolicyRequest(
            String id,
            CapabilityPolicyRecordKind recordKind) { }

    public record GetPolicyResponse(
            Optional<Policy> policy) { }

    public record GetPolicyOperationRequest(
            String operationId) { }

    public record GetPolicyOperationResponse(
            Optional<CapabilityPolicyOperation> receipt) { }

    public record ListPoliciesRequest(
            CapabilityPolicyRecordKind recordKind,
            Optional<PageRequest> page) { }

    public record ListPoliciesResponse(
            List<Policy> policies,
            long catalogGeneration,
            Optional<PageResponse> page) { }

    public record CapabilityInspectionPolicy(
            String id,
            long revision,
            String digest) { }

    public record CapabilityBindingInspection(
            Optional<String> definitionDigest,
            Optional<CapabilityInspectionPolicy> providerBinding,
            List<CapabilityInspectionPolicy> policies,
            String providerProfile,
            String providerConfigurationDigest,
            long providerConfigurationEpoch,
            String state) { }

    public record CapabilityDescriptor(
            String id,
            String contract,
            String provider,
            List<String> operations,
            Map<String, String> attributes,
            Optional<CapabilityBindingInspection> inspection) { }

    public record ListCapabilitiesRequest(
            Optional<String> contractPrefix,
            Optional<String> provider,
            Optional<PageRequest> page,
            String deploymentId,
            boolean includeNodeUsage) { }

    public record CapabilityInspectionRevision(
            String deploymentId,
            String revisionId,
            String componentDigest,
            Optional<String> publicationId,
            long routeGeneration,
            long catalogTransaction) { }

    public record CapabilityResourceUsage(
            String scope,
            Map<String, Long> counters,
            List<String> unavailable) { }

    public record ListCapabilitiesResponse(
            List<CapabilityDescriptor> capabilities,
            Optional<PageResponse> page,
            Optional<CapabilityInspectionRevision> revision,
            Optional<CapabilityResourceUsage> tenantUsage,
            Optional<CapabilityResourceUsage> nodeUsage,
            String state) { }

    public record CapabilityInspectionCeiling(
            int operations,
            long inputBytes,
            long outputBytes,
            long wallTimeMillis) { }

    public record PublicationRef(
            String id,
            String tenant) { }

    public record ReleaseSelector(
            Optional<String> componentDigest,
            Optional<PublicationRef> publication) { }

    public record PublicationIdentity(
            PublicationRef publication,
            String componentDigest,
            String packageDigest) { }

    public record CallOptions(
            Optional<Long> timeoutMillis) { }

    public record RequestIdentity(
            Optional<String> activationId,
            Optional<String> operationId) { }

    public record UnsupportedWireValue(
            String field,
            String value) { }

    public record ResponseMetadata(
            RequestIdentity identity,
            OutcomeKnowledge outcome,
            Optional<AuditAck> auditAck,
            Optional<String> auditStatus,
            Optional<Long> auditAttemptSequence) { }

    public record ClientFailure(
            FailureCategory category,
            String message,
            Optional<Integer> grpcStatus,
            Optional<PlatformError> platformError,
            boolean dispatched,
            OutcomeKnowledge outcome,
            RequestIdentity identity,
            Optional<AuditAck> auditAck,
            Optional<String> auditStatus,
            Optional<UnsupportedWireValue> unsupportedWireValue,
            Optional<Long> auditAttemptSequence) { }

    public record ClientResponse<Response>(Response value, ResponseMetadata metadata) { }

    public interface ClientProfile {
        CompletableFuture<ClientResponse<InvokeResponse>> invoke(
                InvokeRequest request, CallOptions options);

        CompletableFuture<ClientResponse<CancelResponse>> cancel(
                CancelRequest request, CallOptions options);

        CompletableFuture<ClientResponse<ActivationStatus>> getActivation(
                GetActivationRequest request, CallOptions options);

        CompletableFuture<ClientResponse<GetPolicyResponse>> getPolicy(
                GetPolicyRequest request, CallOptions options);

        CompletableFuture<ClientResponse<ListPoliciesResponse>> listPolicies(
                ListPoliciesRequest request, CallOptions options);

        CompletableFuture<ClientResponse<ListCapabilitiesResponse>> listCapabilities(
                ListCapabilitiesRequest request, CallOptions options);

        CompletableFuture<ClientResponse<ApplyPolicyResponse>> applyPolicy(
                ApplyPolicyRequest request, CallOptions options);

        CompletableFuture<ClientResponse<GetPolicyOperationResponse>> getPolicyOperation(
                GetPolicyOperationRequest request, CallOptions options);

    }

    public static final class ClientException extends RuntimeException {
        private static final long serialVersionUID = 1L;
        private final ClientFailure failure;

        public ClientException(ClientFailure failure) {
            super(failure.message());
            this.failure = failure;
        }

        public ClientFailure failure() { return failure; }
    }

    public static final class ClientCancellationException extends java.util.concurrent.CancellationException {
        private static final long serialVersionUID = 1L;
        private final ClientFailure failure;

        public ClientCancellationException(ClientFailure failure) {
            super(failure.message());
            this.failure = failure;
        }

        public ClientFailure failure() { return failure; }
    }

    public static Optional<ClientFailure> clientFailure(Throwable failure) {
        for (int depth = 0; failure != null && depth < 8; depth++) {
            if (failure instanceof ClientException typed) return Optional.of(typed.failure());
            if (failure instanceof ClientCancellationException typed) return Optional.of(typed.failure());
            failure = failure.getCause();
        }
        return Optional.empty();
    }

    public static long parseU64Decimal(String value) {
        if (!value.matches("0|[1-9][0-9]{0,19}")) throw new NumberFormatException("invalid uint64 decimal");
        return Long.parseUnsignedLong(value);
    }

    public static String formatU64Decimal(long value) { return Long.toUnsignedString(value); }
}
